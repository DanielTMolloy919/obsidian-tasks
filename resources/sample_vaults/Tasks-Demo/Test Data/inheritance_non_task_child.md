-  [ ] #task task parent
    - [ ] #task task child
    - [ ] non-task child
