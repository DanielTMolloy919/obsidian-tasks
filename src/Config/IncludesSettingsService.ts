import type { IncludesMap } from './Settings';

export class IncludesSettingsService {
    /**
     * Adds a new include to the map with a unique key
     * @param includes The current includes map (will not be modified)
     * @returns An object with the updated includes map and the new key
     */
    public addInclude(includes: Readonly<IncludesMap>): { includes: IncludesMap; newKey: string } {
        const newKey = this.generateUniqueKey(includes);
        const newIncludes = { ...includes };
        newIncludes[newKey] = '';
        return {
            includes: newIncludes,
            newKey,
        };
    }

    /**
     * Checks if a new key would create a duplicate in the includes map
     * @param includes The current includes map (will not be modified)
     * @param originalKey The original key (to exclude from comparison)
     * @param newKey The new key to check
     * @returns True if the key would be a duplicate, false otherwise
     */
    public isDuplicateKey(includes: Readonly<IncludesMap>, originalKey: string, newKey: string): boolean {
        // Check if it's the same as the original key (after trimming)
        if (originalKey === newKey) {
            return false; // Same key, not a duplicate
        }

        // Check against all existing keys
        for (const existingKey of Object.keys(includes)) {
            // Skip the original key
            if (existingKey !== originalKey && existingKey === newKey) {
                return true; // Found a duplicate
            }
        }

        return false;
    }

    /**
     * Generates a unique key for a new include
     * @param includes The current includes map
     * @returns A unique key string
     */
    private generateUniqueKey(includes: IncludesMap): string {
        const baseKey = 'new_key';
        let suffix = 1;
        while (Object.prototype.hasOwnProperty.call(includes, `${baseKey}_${suffix}`)) {
            suffix++;
        }
        return `${baseKey}_${suffix}`;
    }
}
