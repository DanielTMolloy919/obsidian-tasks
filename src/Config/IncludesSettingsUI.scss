/* ---------------------------------------------- */
/* Includes section                               */
/* ---------------------------------------------- */

.tasks-settings {
    // Main wrapper for includes section
    .tasks-includes-wrapper {
        width: 100%;
        position: relative;
        transition: all 0.2s ease;

        // Drag visual state
        &.tasks-includes-dragging {
            opacity: 0.5;
            transform: rotate(2deg);
        }

        &.tasks-includes-drop-above::before {
            content: '';
            position: absolute;
            top: -2px;
            left: 0;
            right: 0;
            height: 4px;
            background-color: var(--interactive-accent);
            border-radius: 2px;
            z-index: 10;
        }

        &.tasks-includes-drop-below::after {
            content: '';
            position: absolute;
            bottom: -2px;
            left: 0;
            right: 0;
            height: 4px;
            background-color: var(--interactive-accent);
            border-radius: 2px;
            z-index: 10;
        }
    }

    // Individual setting row
    .tasks-includes-setting {
        // Name input field (left column)
        .tasks-includes-key {
            grid-area: key;

            // Error state styling
            &.has-error {
                border-color: var(--text-error);
                border-width: 2px;
            }
        }

        // Value textarea (right column)
        .tasks-includes-value {
            grid-area: value;

            min-width: 300px;
            min-height: 3em;
            font-family: var(--font-monospace);

            // Horizontal-only resizing and scrolling
            resize: horizontal;
            overflow-x: auto;
            overflow-y: hidden;
            white-space: pre;        // Preserve formatting
        }

        .tasks-includes-drag-handle {
            grid-area: drag;
            color: var(--text-muted);

            &:hover {
                color: var(--text-normal);
            }
        }

        .tasks-includes-delete-button {
            grid-area: delete;
        }

        // Control alignment adjustment
        .setting-item-control {
            justify-content: start;
            display: grid;
            grid-template-columns: 200px 1fr auto auto;
            grid-template-areas: "key value drag delete";
            gap: 4px;

            align-items: unset;

            // Put a border around the elements for each Include, to make structure easier
            // to see.
            border: 1px solid var(--background-modifier-border) !important;
            padding: 0.5em !important;
            background-color: var(--background-secondary) !important;
            border-radius: 4px !important;

            // Force wrapping behavior on narrow screens
            @container (max-width: 600px) {
                grid-template-columns: 5fr 1fr 1fr;
                grid-template-areas:
                  "key drag delete"
                  "value value value";

                .tasks-includes-key {
                    // Make the Name input field full-width.
                    // The value textarea will take up most of the width, with the Delete button to its right.
                    width: 100%;
                }
            }

            @container (max-width: 500px) {
                .tasks-includes-value {
                    // Make the Value textarea full-width, forcing the Delete button on to the third row.
                    width: 100%;
                    min-width: 0;
                }
            }
        }
    }
}
