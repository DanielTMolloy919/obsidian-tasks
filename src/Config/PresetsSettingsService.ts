import { renameKeyInRecordPreservingOrder } from '../lib/RecordHelpers';
import type { PresetsMap } from './Settings';

/**
 * Represents a map of include keys and their current values
 * used during validation
 */
export interface RenamesInProgress {
    [originalName: string]: string;
}

export class RenameResult {
    public readonly originalName: string;
    public readonly isValid: boolean;
    public readonly errorMessage: string | null;

    constructor(originalName: string, isValid: boolean, errorMessage: string | null) {
        this.originalName = originalName;
        this.isValid = isValid;
        this.errorMessage = errorMessage;
    }
}

/**
 * Result of validating multiple include values at once
 */
export interface RenameResults {
    [originalName: string]: RenameResult;
}

export class PresetsSettingsService {
    /**
     * Validates multiple include names against each other
     * @param renames Map of original keys to their current values in UI
     * @returns Object mapping each key to its validation result
     */
    public validateRenames(renames: RenamesInProgress): RenameResults {
        const results: RenameResults = {};

        // Check each key against all others
        for (const [originalName, newName] of Object.entries(renames)) {
            const includesWithOtherPendingRenames: PresetsMap = {};

            for (const [otherOriginalName, otherNewName] of Object.entries(renames)) {
                // Skip the name being validated to avoid false duplicate matches.
                if (otherOriginalName !== originalName) {
                    includesWithOtherPendingRenames[otherNewName] = '';
                }
            }

            // Pass empty string as keyBeingRenamed since we're not excluding any key from duplicate check
            results[originalName] = this.validateRename(includesWithOtherPendingRenames, '', newName);
        }

        return results;
    }

    /**
     * Validates if an include name is valid
     * @param includes The current presets map
     * @param keyBeingRenamed The key being renamed
     * @param newName The proposed name to validate
     * @returns An object with validation result and error message if any
     */
    public validateRename(includes: Readonly<PresetsMap>, keyBeingRenamed: string, newName: string): RenameResult {
        // Check for empty name
        if (!newName || newName.trim() === '') {
            return new RenameResult(keyBeingRenamed, false, 'Include name cannot be empty or all whitespace');
        }

        for (const existingKey of Object.keys(includes)) {
            // Skip the key being renamed
            if (existingKey === keyBeingRenamed) {
                continue;
            }

            if (existingKey.trim() === newName.trim()) {
                return new RenameResult(keyBeingRenamed, false, 'An include with this name already exists');
            }
        }

        return new RenameResult(keyBeingRenamed, true, null);
    }

    /**
     * Adds a new include to the map with a unique key
     * @param includes The current presets map (will not be modified)
     * @returns An object with the updated presets map and the new key
     */
    public addPreset(includes: Readonly<PresetsMap>): { includes: PresetsMap; newKey: string } {
        const newKey = this.generateUniqueKey(includes);
        const newIncludes = { ...includes };
        newIncludes[newKey] = '';
        return {
            includes: newIncludes,
            newKey,
        };
    }

    /**
     * Renames a key in the presets map, preserving order
     * @param includes The current presets map (will not be modified)
     * @param keyBeingRenamed The existing key that would be renamed
     * @param proposedNewName The new name being considered
     * @returns The updated presets map, or null if the operation failed (for example, duplicate key)
     */
    public renamePreset(
        includes: Readonly<PresetsMap>,
        keyBeingRenamed: string,
        proposedNewName: string,
    ): PresetsMap | null {
        // Validate inputs
        if (!proposedNewName || proposedNewName.trim() === '') {
            return null; // Empty keys are not allowed
        }

        proposedNewName = proposedNewName.trim();

        // Check if this would create a duplicate
        if (this.wouldCreateDuplicateKey(includes, keyBeingRenamed, proposedNewName)) {
            return null;
        }

        return renameKeyInRecordPreservingOrder(includes, keyBeingRenamed, proposedNewName);
    }

    /**
     * Deletes an include from the map
     * @param includes The current presets map (will not be modified)
     * @param key The key to delete
     * @returns The updated presets map
     */
    public deletePreset(includes: Readonly<PresetsMap>, key: string): PresetsMap {
        const newIncludes = { ...includes };
        delete newIncludes[key];
        return newIncludes;
    }

    /**
     * Updates the value of an include
     * @param includes The current presets map (will not be modified)
     * @param key The key to update
     * @param value The new value
     * @returns The updated presets map
     */
    public updatePresetValue(includes: Readonly<PresetsMap>, key: string, value: string): PresetsMap {
        const newIncludes = { ...includes };
        newIncludes[key] = value;
        return newIncludes;
    }

    /**
     * Checks if renaming a key would create a duplicate in the presets map
     * @param includes The presets map to check against
     * @param keyBeingRenamed The existing key that would be renamed
     * @param proposedNewName The new name being considered
     * @returns True if the proposed new name would conflict with an existing key
     */
    public wouldCreateDuplicateKey(
        includes: Readonly<PresetsMap>,
        keyBeingRenamed: string,
        proposedNewName: string,
    ): boolean {
        // Normalize the proposedNewName once
        const normalizedNewName = proposedNewName.trim();

        // If it's the same key (after trimming), it's not a duplicate
        if (keyBeingRenamed.trim() === normalizedNewName) {
            return false;
        }

        // Check against all existing keys
        for (const existingKey of Object.keys(includes)) {
            // Skip the key being renamed (exact reference check)
            if (existingKey !== keyBeingRenamed && existingKey.trim() === normalizedNewName) {
                return true; // Found a duplicate
            }
        }

        return false;
    }

    /**
     * Generates a unique key for a new include
     * @param includes The current presets map
     * @returns A unique key string
     */
    private generateUniqueKey(includes: Readonly<PresetsMap>): string {
        const baseKey = 'new_key';
        let suffix = 1;
        while (Object.prototype.hasOwnProperty.call(includes, `${baseKey}_${suffix}`)) {
            suffix++;
        }
        return `${baseKey}_${suffix}`;
    }

    /**
     * Reorders an include to a specific position in the map
     * @param includes The current presets map (will not be modified)
     * @param key The key to reorder
     * @param newIndex The target position (0-based index)
     * @returns The updated presets map, or null if the operation failed
     */
    public reorderPreset(includes: Readonly<PresetsMap>, key: string, newIndex: number): PresetsMap | null {
        const keys = Object.keys(includes);
        const currentIndex = keys.indexOf(key);

        // Validate inputs
        if (currentIndex === -1) {
            return null; // Key doesn't exist
        }

        if (newIndex < 0 || newIndex >= keys.length) {
            return null; // Invalid target position
        }

        if (currentIndex === newIndex) {
            // No change needed, but return a copy
            return { ...includes };
        }

        // Create new key order
        const newKeys = [...keys];

        // Remove the key from its current position
        newKeys.splice(currentIndex, 1);

        // Insert it at the new position
        newKeys.splice(newIndex, 0, key);

        // Rebuild the map in the new order
        const newIncludes: PresetsMap = {};
        for (const k of newKeys) {
            newIncludes[k] = includes[k];
        }

        return newIncludes;
    }
}
