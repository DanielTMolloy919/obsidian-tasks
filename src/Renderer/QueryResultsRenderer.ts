import { MarkdownRenderChild } from 'obsidian';

export class QueryResultsRenderer extends MarkdownRenderChild {
    constructor(container: HTMLElement) {
        super(container);
    }
}
