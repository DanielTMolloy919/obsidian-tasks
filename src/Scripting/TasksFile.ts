import type { CachedMetadata } from 'obsidian';

/**
 * A simple class to provide access to file information via 'task.file' in scripting code.
 */
export class TasksFile {
    private readonly _path: string;
    private readonly _cachedMetadata: CachedMetadata;

    constructor(path: string, cachedMetadata: CachedMetadata = {}) {
        this._path = path;
        this._cachedMetadata = cachedMetadata;
    }

    /**
     * Return the path to the file.
     */
    get path(): string {
        return this._path;
    }

    /**
     * Return Obsidian's [CachedMetadata](https://docs.obsidian.md/Reference/TypeScript+API/CachedMetadata)
     * for this file, if available.
     *
     * @note This is currently only populated for Task objects when read in the Obsidian plugin.
     *       It's not populated for queries in the plugin, nor in most unit tests.
     */
    public get cachedMetadata(): CachedMetadata {
        return this._cachedMetadata;
    }

    /**
     * Return the path to the file, with the filename extension removed.
     */
    get pathWithoutExtension(): string {
        return this.withoutExtension(this.path);
    }

    private withoutExtension(value: string) {
        return value.replace(/\.md$/, '');
    }

    /**
     * Return the root to the file.
     */
    get root(): string {
        let path = this.path.replace(/\\/g, '/');

        if (path.charAt(0) === '/') {
            path = path.substring(1);
        }

        const separatorIndex = path.indexOf('/');
        if (separatorIndex == -1) {
            return '/';
        }
        return path.substring(0, separatorIndex + 1);
    }

    get folder(): string {
        const path = this.path;
        const fileNameWithExtension = this.filename;
        const folder = path.substring(0, path.lastIndexOf(fileNameWithExtension));
        if (folder === '') {
            return '/';
        }
        return folder;
    }

    /**
     * Return the filename including the extension.
     */
    get filename(): string {
        // Copied from Task.filename and FilenameField.value() initially
        const fileNameMatch = this.path.match(/([^/]+)$/);
        if (fileNameMatch !== null) {
            return fileNameMatch[1];
        } else {
            return '';
        }
    }

    get filenameWithoutExtension(): string {
        return this.withoutExtension(this.filename);
    }
}
