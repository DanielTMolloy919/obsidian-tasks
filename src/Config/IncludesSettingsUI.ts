import { Setting, TextAreaComponent } from 'obsidian';
import type TasksPlugin from '../main';
import { IncludesSettingsService } from './IncludesSettingsService';
import { type IncludesMap, type Settings, updateSettings } from './Settings';

export type RefreshViewCallback = () => void;

export class IncludesSettingsUI {
    private readonly plugin: TasksPlugin;
    public readonly includesSettingsService = new IncludesSettingsService();

    constructor(plugin: TasksPlugin) {
        this.plugin = plugin;
    }

    public setupAutoResizingTextarea(textArea: TextAreaComponent) {
        const resize = () => {
            textArea.inputEl.style.height = 'auto'; // reset first
            textArea.inputEl.style.height = `${textArea.inputEl.scrollHeight}px`;
        };

        // Initial resize
        resize();

        // Resize on input
        textArea.inputEl.addEventListener('input', resize);
    }

    createAddNewIncludeButton(containerEl: HTMLElement, settings: Settings, renderIncludes: RefreshViewCallback) {
        new Setting(containerEl).addButton((btn) => {
            btn.setButtonText('Add new include')
                .setCta()
                .onClick(async () => {
                    const { includes: updatedIncludes } = this.includesSettingsService.addInclude(settings.includes);
                    await this.saveIncludesSettings(updatedIncludes, settings, renderIncludes);
                });
        });
    }

    /**
     * Updates settings with new includes and refreshes UI if needed
     * @param updatedIncludes The new includes map
     * @param settings The current settings object to update
     * @param refreshView Callback to refresh the view (pass null if no refresh is needed)
     */
    public async saveIncludesSettings(
        updatedIncludes: IncludesMap,
        settings: Settings,
        refreshView: RefreshViewCallback | null,
    ) {
        // Update the settings in storage
        updateSettings({ includes: updatedIncludes });
        await this.plugin.saveSettings();

        // Update the local settings object to reflect the changes
        settings.includes = { ...updatedIncludes };

        // Refresh the view if a callback was provided
        if (refreshView) {
            refreshView();
        }
    }
}

export function renderIncludeItem(
    includesSettingsUI: any,
    includesContainer: HTMLDivElement,
    settings: Settings,
    key: string,
    value: string,
    renderIncludes: RefreshViewCallback,
) {
    const wrapper = includesContainer.createDiv({ cls: 'tasks-includes-wrapper' });
    const setting = new Setting(wrapper);
    setting.settingEl.addClass('tasks-includes-setting');

    // Add name input field
    setting.addText((text) => {
        text.setPlaceholder('Name').setValue(key);
        text.inputEl.addClass('tasks-includes-key');

        let newKey = key;

        text.inputEl.addEventListener('input', (e) => {
            newKey = (e.target as HTMLInputElement).value;
        });

        // Handle renaming an include
        const commitRename = async () => {
            if (newKey && newKey !== key) {
                const updatedIncludes = includesSettingsUI.includesSettingsService.renameInclude(
                    settings.includes,
                    key,
                    newKey,
                );
                if (updatedIncludes) {
                    await includesSettingsUI.saveIncludesSettings(updatedIncludes, settings, renderIncludes);
                }
            }
        };

        text.inputEl.addEventListener('blur', commitRename);
        text.inputEl.addEventListener('keydown', async (e) => {
            if (e.key === 'Enter') {
                e.preventDefault();
                text.inputEl.blur(); // trigger blur handler
            }
        });
    });

    // Add value textarea
    setting.addTextArea((textArea) => {
        textArea.inputEl.addClass('tasks-includes-value');
        textArea.setPlaceholder('Query or filter text...').setValue(value);

        includesSettingsUI.setupAutoResizingTextarea(textArea);

        return textArea.onChange(async (newValue) => {
            const updatedIncludes = includesSettingsUI.includesSettingsService.updateIncludeValue(
                settings.includes,
                key,
                newValue,
            );
            await includesSettingsUI.saveIncludesSettings(updatedIncludes, settings, null);
        });
    });

    // Add delete button
    setting.addExtraButton((btn) => {
        btn.setIcon('cross')
            .setTooltip('Delete')
            .onClick(async () => {
                const updatedIncludes = includesSettingsUI.includesSettingsService.deleteInclude(
                    settings.includes,
                    key,
                );
                await includesSettingsUI.saveIncludesSettings(updatedIncludes, settings, renderIncludes);
            });
    });
}
