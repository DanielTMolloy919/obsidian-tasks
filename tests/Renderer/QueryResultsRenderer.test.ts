/**
 * @jest-environment jsdom
 */
import moment from 'moment';
import type { Task } from 'Task/Task';
import { GlobalFilter } from '../../src/Config/GlobalFilter';
import { State } from '../../src/Obsidian/Cache';
import { QueryResultsRenderer } from '../../src/Renderer/QueryResultsRenderer';
import { TasksFile } from '../../src/Scripting/TasksFile';
import inheritance_non_task_child from '../Obsidian/__test_data__/inheritance_non_task_child.json';
import inheritance_rendering_sample from '../Obsidian/__test_data__/inheritance_rendering_sample.json';
import inheritance_task_2listitem_3task from '../Obsidian/__test_data__/inheritance_task_2listitem_3task.json';
import internal_heading_links_test from '../Obsidian/__test_data__/internal_heading_links.json';
import { readTasksFromSimulatedFile } from '../Obsidian/SimulatedFile';
import { verifyWithFileExtension } from '../TestingTools/ApprovalTestHelpers';
import { prettifyHTML } from '../TestingTools/HTMLHelpers';
import { TaskBuilder } from '../TestingTools/TaskBuilder';
import { toMarkdown } from '../TestingTools/TestHelpers';
import { mockHTMLRenderer } from './RenderingTestHelpers';

window.moment = moment;

beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date('2023-07-05'));
});

afterEach(() => {
    jest.useRealTimers();
    GlobalFilter.getInstance().reset();
});

function makeQueryResultsRenderer(source: string, tasksFile: TasksFile) {
    return new QueryResultsRenderer(
        'block-language-tasks',
        source,
        tasksFile,
        () => Promise.resolve(),
        null,
        mockHTMLRenderer,
    );
}

describe('QueryResultsRenderer tests', () => {
    async function verifyRenderedTasksHTML(allTasks: Task[], source: string = '') {
        const renderer = makeQueryResultsRenderer(source, new TasksFile('query.md'));
        const queryRendererParameters = {
            allTasks,
            allMarkdownFiles: [],
            backlinksClickHandler: () => Promise.resolve(),
            backlinksMousedownHandler: () => Promise.resolve(),
            editTaskPencilClickHandler: () => Promise.resolve(),
        };
        const container = document.createElement('div');

        await renderer.render(State.Warm, allTasks, container, queryRendererParameters);

        const taskAsMarkdown = `<!--
${toMarkdown(allTasks)}
-->\n\n`;

        const prettyHTML = prettifyHTML(container.outerHTML);
        verifyWithFileExtension(taskAsMarkdown + prettyHTML, 'html');
    }

    it('fully populated task', async () => {
        const allTasks = [TaskBuilder.createFullyPopulatedTask()];
        await verifyRenderedTasksHTML(allTasks, 'show urgency');
    });

    it('fully populated task - short mode', async () => {
        const allTasks = [TaskBuilder.createFullyPopulatedTask()];
        await verifyRenderedTasksHTML(allTasks, 'show urgency\nshort mode');
    });

    const showTree = 'show tree\n';
    const hideTree = 'hide tree\n';

    it('parent-child items hidden', async () => {
        const allTasks = readTasksFromSimulatedFile(inheritance_rendering_sample);
        await verifyRenderedTasksHTML(allTasks, hideTree + 'sort by function task.lineNumber');
    });

    it('parent-child items', async () => {
        const allTasks = readTasksFromSimulatedFile(inheritance_rendering_sample);
        await verifyRenderedTasksHTML(allTasks, showTree + 'sort by function task.lineNumber');
    });

    it('parent-child items reverse sorted', async () => {
        const allTasks = readTasksFromSimulatedFile(inheritance_rendering_sample);
        await verifyRenderedTasksHTML(allTasks, showTree + 'sort by function reverse task.lineNumber');
    });

    it('should render tasks without their parents', async () => {
        // example chosen to match subtasks whose parents do not match the query
        const allTasks = readTasksFromSimulatedFile(inheritance_task_2listitem_3task);
        await verifyRenderedTasksHTML(allTasks, showTree + 'description includes grandchild');
    });

    it('should render non task check box when global filter is enabled', async () => {
        GlobalFilter.getInstance().set('#task');
        const allTasks = readTasksFromSimulatedFile(inheritance_non_task_child);
        await verifyRenderedTasksHTML(allTasks, showTree);
    });
});

describe('QueryResultsRenderer - responding to file edits', () => {
    it('should update the query its file path is changed', () => {
        // Arrange
        const source = 'path includes {{query.file.path}}';
        const renderer = makeQueryResultsRenderer(source, new TasksFile('oldPath.md'));
        expect(renderer.query.explainQuery()).toContain('path includes oldPath.md');

        // Act
        renderer.setTasksFile(new TasksFile('newPath.md'));

        // Assert
        expect(renderer.query.explainQuery()).toContain('path includes newPath.md');
    });
});

describe('QueryResultsRenderer - internal heading links', () => {
    let allTasks: Task[];

    beforeAll(() => {
        allTasks = readTasksFromSimulatedFile(internal_heading_links_test);
    });

    async function testHeadingLinks(taskDescription: string, queryFilePath: string, expectedLinks: string[]) {
        const task = allTasks.find((t) => t.description.includes(taskDescription));
        if (!task) throw new Error(`Task with description "${taskDescription}" not found`);

        const renderer = makeQueryResultsRenderer('', new TasksFile(queryFilePath));
        const queryRendererParameters = {
            allTasks: [task],
            allMarkdownFiles: [],
            backlinksClickHandler: () => Promise.resolve(),
            backlinksMousedownHandler: () => Promise.resolve(),
            editTaskPencilClickHandler: () => Promise.resolve(),
        };
        const container = document.createElement('div');

        await renderer.render(State.Warm, [task], container, queryRendererParameters);

        const renderedText = container.innerHTML;
        expectedLinks.forEach((link) => expect(renderedText).toContain(link));
    }

    it('should not modify internal heading links when rendering in same file', async () => {
        await testHeadingLinks('[[#Basic Internal Links]]', 'original.md', [
            '[[Test Data/internal_heading_links.md#Basic Internal Links|Basic Internal Links]]',
        ]);
    });

    it('should convert internal heading links when rendering in different file', async () => {
        await testHeadingLinks('[[#Basic Internal Links]]', 'query.md', [
            '[[Test Data/internal_heading_links.md#Basic Internal Links|Basic Internal Links]]',
        ]);
    });

    it('should handle multiple internal heading links in one description', async () => {
        await testHeadingLinks('[[#Multiple Links In One Task]] and [[#Simple Headers]]', 'query.md', [
            '[[Test Data/internal_heading_links.md#Multiple Links In One Task|Multiple Links In One Task]]',
            '[[Test Data/internal_heading_links.md#Simple Headers|Simple Headers]]',
        ]);
    });

    it('should not modify regular file links', async () => {
        await testHeadingLinks('[[Other File]]', 'query.md', ['[[Other File]]']);
    });

    it('should handle mixed link types correctly', async () => {
        await testHeadingLinks('[[Other File]] and [[#Mixed Link Types]]', 'query.md', [
            '[[Other File]]',
            '[[Test Data/internal_heading_links.md#Mixed Link Types|Mixed Link Types]]',
        ]);
    });

    it('should handle links with spaces in headers', async () => {
        await testHeadingLinks('[[#Headers With Spaces]]', 'query.md', [
            '[[Test Data/internal_heading_links.md#Headers With Spaces|Headers With Spaces]]',
        ]);
    });

    it('should handle special characters in headers', async () => {
        await testHeadingLinks('[[#Headers With Dashes]]', 'query.md', [
            '[[Test Data/internal_heading_links.md#Headers With Dashes|Headers With Dashes]]',
        ]);
        await testHeadingLinks('[[#Headers With Underscores]]', 'query.md', [
            '[[Test Data/internal_heading_links.md#Headers With Underscores|Headers With Underscores]]',
        ]);
    });

    it('should handle links with aliases', async () => {
        await testHeadingLinks('[[#Aliased Links|I am an alias]]', 'query.md', [
            '[[Test Data/internal_heading_links.md#Aliased Links|I am an alias]]',
        ]);
    });

    it('should not modify formatted text that looks like links in code blocks', async () => {
        await testHeadingLinks('`[[#Links In Code Blocks]]`', 'query.md', ['`[[#Links In Code Blocks]]`']);
    });

    it('should not modify escaped links', async () => {
        await testHeadingLinks('\\[\\[#Escaped Links\\]\\]', 'query.md', ['\\[\\[#Escaped Links\\]\\]']);
    });
});
