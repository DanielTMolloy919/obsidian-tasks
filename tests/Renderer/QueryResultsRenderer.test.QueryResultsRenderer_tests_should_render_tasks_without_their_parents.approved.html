<div>
  <ul class="contains-task-list plugin-tasks-query-result tasks-layout-hide-urgency"></ul>
  <div class="task-count">3 tasks</div>
</div>
